// gulp-iife/lib/iife.js
"use strict";

var _ = require("lodash");

module.exports = {
    surround: surround
};

var defaultOptions = {
    args: undefined,
    params: undefined,
    prependSemicolon: true,
    useStrict: true,
    trimCode: true
};

function surround(code, userOptions) {
    var options = _.merge({}, defaultOptions, userOptions);

    var useStrictLines = options.useStrict ? ["\"use strict\";", ""] : [];
    var trimmedCode = options.trimCode ? code.trim() : code;
    var prependedSemicolon = options.prependSemicolon ? ";" : "";
    var bindThis = options.bindThis ? ".bind(this)" : "";

    var _getArgsAndParams = getArgsAndParams(options);

    var args = _getArgsAndParams.args;
    var params = _getArgsAndParams.params;

    var lines = [prependedSemicolon + "(function(" + params + ") {"].concat(useStrictLines, [trimmedCode, "}" + bindThis + "(" + args + "));", ""]);

    return lines.join("\n");
}

function getArgsAndParams(options) {
    var params = options.params || options.args || [];
    var args = options.args || options.params || [];

    return {
        args: args.join(", "),
        params: params.join(", ")
    };
}
